### PR TITLE
It's time to add some test for the installer itself

### PR DESCRIPTION
**Pull request by xtreamwayz**
It tests: 
- configuation 
- containers
- routers
- template renderers 
- the homepage action for each test where possible

### CODE REVIEW

File: src/ExpressiveInstaller/OptionalPackages.php [hunk 1]
```diff
@@ -42,27 +42,27 @@ class OptionalPackages
     /**
      * @var array
      */
-    private static $config;
+    public static $config;
```
Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: Why public?

If it's for testing, the various `assertAttribute*()` assertions can do assertions on non-public properties. That said, I've not tried on static properties; if those present an issue, I'd argue reflection is still a better way to test these.

File: src/ExpressiveInstaller/OptionalPackages.php [hunk 2]
```diff
@@ -199,11 +210,13 @@ public static function install(Event $event)
      * this one) and assets (including configuration and templates).
      *
      * @param IOInterface $io
+     * @param string      $projectRoot
      */
-    private static function cleanUp(IOInterface $io)
+    private static function cleanUp(IOInterface $io, $projectRoot)
     {
-        $io->write("<info>Removing Expressive installer classes and configuration</info>");
+        $io->write("<info>Removing Expressive installer classes, configuration and tests</info>");
```
Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: Use an oxford comma here: "classes, configuration, and tests".

File: src/ExpressiveInstaller/OptionalPackages.php [hunk 3]
```diff
@@ -288,7 +301,7 @@ private static function askQuestion(Composer $composer, IOInterface $io, $questi
      * @param $packageName
      * @param $packageVersion
      */
-    private static function addPackage(IOInterface $io, $packageName, $packageVersion)
+    public static function addPackage(IOInterface $io, $packageName, $packageVersion)
```
Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: Why public?

File: src/ExpressiveInstaller/OptionalPackages.php [hunk 4]
```diff
@@ -340,7 +353,7 @@ private static function addPackage(IOInterface $io, $packageName, $packageVersio
      * @param string $target Destination.
      * @param bool $force whether or not to copy over an existing file.
      */
-    private static function copyFile(IOInterface $io, $projectRoot, $source, $target, $force = false)
+    public static function copyFile(IOInterface $io, $projectRoot, $source, $target, $force = false)
```
Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: Why public?

File: src/ExpressiveInstaller/OptionalPackages.php (hunk 1 above)
Comment (xtreamwayz) [inline src/ExpressiveInstaller/OptionalPackages.php]: Those are not being tested. They are public so I can set them from within a test. Doing it this way I don't need to repeat the code to setup the OptionalPackages class.

File: test/ExpressiveInstallerTest/ContainersTest.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        // $containerOption, $routerOption, $copyFilesKey, $expectedResponseStatusCode, $expectedContainer
+        return [
+            [1, 2, 'copy-files', 200, AuraContainer::class],
+            [2, 2, 'copy-files', 200, PimpleInteropContainer::class],
+            [3, 2, 'copy-files', 200, ZendServiceManagerContainer::class],
+
+            [1, 2, 'minimal-files', 404, AuraContainer::class],
+            [2, 2, 'minimal-files', 404, PimpleInteropContainer::class],
+            [3, 2, 'minimal-files', 404, ZendServiceManagerContainer::class],
+        ];
```
Comment (weierophinney) [inline test/ExpressiveInstallerTest/ContainersTest.php]: Please add keys to each, detailing what they are testing; this makes understanding which failed simpler:

```php
return [
    'aura-full-install' => [ /* ... */ ],
    'aura-minimal' => [/* ... */ ],
    /* etc. */
];
````

File: test/ExpressiveInstallerTest/InstallerTestCase.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    protected $teardownFiles = [];
+
+    public function setup()
+    {
+        $this->response = null;
+
+        $this->cleanup();
+
+        $this->io = $this->prophesize('Composer\IO\IOInterface')->reveal();
```
Comment (weierophinney) [inline test/ExpressiveInstallerTest/InstallerTestCase.php]: Shouldn't we wait to reveal this until later, so we can do assertions on what was written?

File: src/ExpressiveInstaller/OptionalPackages.php (hunk 3 above)
Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: nm; I see why (you call on these in the tests)

File: test/ExpressiveInstallerTest/RoutersTest.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        // $containerOption, $routerOption, $copyFilesKey, $expectedResponseStatusCode, $expectedRoutes, $expectedRouter
+        return [
+            [3, 1, 'copy-files', 200, $this->expectedRoutes, Router\AuraRouter::class],
+            [3, 2, 'copy-files', 200, $this->expectedRoutes, Router\FastRouteRouter::class],
+            [3, 3, 'copy-files', 200, $this->expectedRoutes, Router\ZendRouter::class],
+
+            [3, 1, 'minimal-files', 404, [], Router\AuraRouter::class],
+            [3, 2, 'minimal-files', 404, [], Router\FastRouteRouter::class],
+            [3, 3, 'minimal-files', 404, [], Router\ZendRouter::class],
+        ];
```
Comment (weierophinney) [inline test/ExpressiveInstallerTest/RoutersTest.php]: Same comment as above (on the containers test): add keys to each test case.

File: test/ExpressiveInstallerTest/TemplateRenderersTest.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        // $expectedTemplateRenderer
+        return [
+            [3, 2, 1, 'copy-files', 200, Expressive\Plates\PlatesRenderer::class],
+            [3, 2, 2, 'copy-files', 200, Expressive\Twig\TwigRenderer::class],
+            [3, 2, 3, 'copy-files', 200, Expressive\ZendView\ZendViewRenderer::class],
+
+            [3, 2, 1, 'minimal-files', 404, Expressive\Plates\PlatesRenderer::class],
+            [3, 2, 2, 'minimal-files', 404, Expressive\Twig\TwigRenderer::class],
+            [3, 2, 3, 'minimal-files', 404, Expressive\ZendView\ZendViewRenderer::class],
+        ];
```
Comment (weierophinney) [inline test/ExpressiveInstallerTest/TemplateRenderersTest.php]: Same comment as above: add keys to each test case.

File: src/ExpressiveInstaller/OptionalPackages.php (hunk 1 above)
Comment (xtreamwayz) [inline src/ExpressiveInstaller/OptionalPackages.php]: https://github.com/zendframework/zend-expressive-skeleton/pull/48/files#diff-733c8720c49c759a0935fc5596c19e72R49

Comment (weierophinney) [inline src/ExpressiveInstaller/OptionalPackages.php]: Couldn't these assignments be done with the Reflection API instead? That way we protect encapsulation for normal consumption.

File: src/ExpressiveInstaller/OptionalPackages.php (hunk 4 above)
Comment (xtreamwayz) [inline src/ExpressiveInstaller/OptionalPackages.php]: Same, it's called in the InstallerTestCase

File: src/ExpressiveInstaller/OptionalPackages.php (hunk 1 above)
Comment (xtreamwayz) [inline src/ExpressiveInstaller/OptionalPackages.php]: I'm not sure how this would work with static classes.

**Pull request merged**